xrootd/client/client.go

```diff
@@ -0,0 +1,141 @@
+// license that can be found in the LICENSE file.
+
+package client // import "go-hep.org/x/hep/xrootd/client"
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/mux"
+	"go-hep.org/x/hep/xrootd/protocol"
+)
+
+// A Client to xrootd server which allows to send requests and receive responses.
+// Concurrent requests are supported.
+// Zero value is invalid, Client should be instantiated using NewClient.
+type Client struct {
+	conn            net.Conn
+	mux             *mux.Mux
+	protocolVersion int32
+}
+
+const responseHeaderSize = 2 + 2 + 4
+
+type responseHeader struct {
+	StreamID   protocol.StreamID
+	Status     uint16
+	DataLength int32
+}
+
+// NewClient creates a client to xrootd server at address.
+// ctx defines a lifetime of the client, once it is done response handling is stopped.
+func NewClient(ctx context.Context, address string) (*Client, error) {
+	conn, err := net.Dial("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{conn, mux.New(), 0}
+
+	go client.consume(ctx)
+
+	if err := client.handshake(ctx); err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (client *Client) consume(ctx context.Context) {
+	var header = &responseHeader{}
+	var headerBytes = make([]byte, responseHeaderSize)
+
+	for {
+		if _, err := io.ReadFull(client.conn, headerBytes); err != nil {
+			// TODO: handle EOF by redirection as specified at http://xrootd.org/doc/dev45/XRdv310.pdf, page 11
+		}
+
+		if err := encoder.Unmarshal(headerBytes, header); err != nil {
+			// TODO: should redirect in case if is not possible to decode a header as well?
+		}
+
+		resp := mux.ServerResponse{make([]byte, header.DataLength), nil}
+		if _, err := io.ReadFull(client.conn, resp.Data); err != nil {
+			resp.Err = err
+		} else if header.Status != protocol.Ok {
+			resp.Err = extractError(header, resp.Data)
+		}
+
+		if err := client.mux.SendData(header.StreamID, resp); err != nil {
+			// TODO: should we just ignore responses to unclaimed stream IDs?
+		}
+
+		if header.Status != protocol.OkSoFar {
+			client.mux.Unclaim(header.StreamID)
+		}
+
+		select {
+		case <-ctx.Done():
+			// TODO: wait for active requests to be completed and handle errors from client.conn.Close().
+			client.conn.Close()
+			client.mux.Close()
+			return
+		default:
+		}
+	}
+}
+
+func extractError(header *responseHeader, data []byte) error {
+	if header.Status == protocol.Error {
+		code := int32(binary.BigEndian.Uint32(data[0:4]))
+		message := string(data[4 : len(data)-1]) // Skip \0 character at the end
+
+		return protocol.ServerError{code, message}
+	}
+	return nil
+}
+
+func (client *Client) callWithBytesAndResponseChannel(ctx context.Context, responseChannel mux.DataRecvChan, request []byte) ([]byte, error) {
+	if _, err := client.conn.Write(request); err != nil {
+		return nil, err
+	}
+
+	more := true
+	var serverResponse mux.ServerResponse
```

Comment:
in the context of the client, I don't think there's an ambiguity as to from which/whom a response is (ie: the server)
I'd be tempted to just call this `resp` or `response`:
```go
var resp mux.ServerResponse
```